### ISSUE TITLE
Mem BIO does not add error queue element on unsupported CTR; fix would break SSL state machine

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#10955]
When doing a test due to a review comment on the new generalized HTTP client https://github.com/openssl/openssl/pull/10667#discussion_r370666642
it turned out that when a memory BIO receives CTRL commandsthat it cannot handle like `BIO_C_DO_STATE_MACHINE` it returns `0` without adding a respective entry in the error queue such that can be pretty hard to find out what went wrong.

When doing the obvious fix, namely to add
```
        BIOerr(0, BIO_R_UNSUPPORTED_METHOD);
```
in the `default:` case of `mem_ctrl()` in `crypto/bio/bss_mem.c` a number of SSL test cases start failing!
```
70-test_asyncio.t
70-test_bad_dtls.t
70-test_recordlen.t
70-test_servername.t
80-test_ssl_new.t
80-test_ssl_old.t
80-test_sslcorrupt.t
```

There must be something pretty fishy going on in the SSL state machine.
My guess is that this is related to https://github.com/openssl/openssl/issues/10021 and an observation made in https://github.com/openssl/openssl/pull/8716#issuecomment-481746415: 
> `SSL_get_error` is sensitive to whether there are errors in the error queue.

**Comment by levitte** [comment]
Yup, we do have spots a little here and there that are incredibly sensitive to the exact message at the top or bottom of the stack.  We definitely need something (have needed for a very long time) that differs between ignorable (non-fatal?) and non-ignorable (fatal?) errors.  We do use -2 as a generic return code for unsupported stuff, but opinions differ regarding how appropriate it is with multi-state returns.  We also have a fatal bit in the generic error codes (see err.h [*]), but that doesn't seem to be looked at anywhere, so I assume that's an idea from longago that has lingered.

-----

[*]
https://github.com/openssl/openssl/blob/2ee4a50ab92020dc49383d5aa644397edac4a59a/include/openssl/err.h#L225-L236

That bit goes far back that I don't have the historical knowledge.  @t-j-h might be able to fill us in.

**Comment by nhorman** [comment]
This seems like a reasonable thing to try fix.

@DDvO you don't happen to have the verbose output from those failing tests available do you, just to help track down where things went off the rails?

**Comment by DDvO** [comment]
The output is straightforward to produce: 
just add `BIOerr(0, BIO_R_UNSUPPORTED_METHOD);` as mentioned in the OP,
and then a hell lot of test cases fail.

Output is attached.
[test_BIO_R_UNSUPPORTED_METHOD.log.txt](https://github.com/user-attachments/files/15846586/test_BIO_R_UNSUPPORTED_METHOD.log.txt)
